### PR TITLE
Who is listening Icon changes

### PR DESCRIPTION
**Pull request by djnunez-aot**
*Issue #142 :*
https://github.com/bcgov/met-public/issues/142

-edit icons

-edit goes to add contact window but changes text to edit contact with current data filled

-delete icon removes card from list but no contact from database

-add react dnd library and implementing dragging for contact components

-update contact form validation

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the met-public license (Apache 2.0).

### CODE REVIEW

File: met-web/src/components/common/ControlledInputComponents/ControlledTextField.tsx [hunk 1]
```diff
@@ -6,17 +6,17 @@ type IFormInputProps = {
     name: string;
 } & TextFieldProps;
 
-const ControlledTextField: FC<IFormInputProps> = ({ name, ...otherProps }) => {
+const ControlledTextField: FC<IFormInputProps> = ({ name, defaultValue, ...otherProps }) => {
     const {
         control,
-        formState: { errors, defaultValues },
```
Comment (jadmsaadaot) [inline met-web/src/components/common/ControlledInputComponents/ControlledTextField.tsx:12]: This works, but react hook form has already a way to manage default values via the defaultValues prop here. The way to use is to revert the changes in this file and to pass {whatever, defaultValues: {...}} to the useForm hook in the AddContactDrawer

**Review by jadmsaadaot — CHANGES_REQUESTED** [review]
Let's leverage React Hook Form formState, and pass defaultValues to userForm to set the default values instead of passing them one by one

File: met-web/src/components/common/ControlledInputComponents/ControlledTextField.tsx (hunk 1 above)
Comment (djnunez-aot) [reply met-web/src/components/common/ControlledInputComponents/ControlledTextField.tsx]: I did the changes requested but it doesnt not seem to be updating correctly

File: met-api/src/met_api/resources/contact.py [hunk 2]
```diff
@@ -88,3 +88,26 @@ def get():
             return ActionResult.success(result=widgets)
         except (KeyError, ValueError) as err:
             return ActionResult.error(str(err))
+
+
+@cors_preflight('PUT')
+@API.route('/')
+class Contact(Resource):
```
Comment (tiagograf-AOT) [inline met-api/src/met_api/resources/contact.py:95]: you don't need to create a new class for this, just use the first one on this file (you can add the 'PUT' allowed method if needed), also I noticed the class names are wrong (submission/survey stuff) rename them to Contact/Contacts

File: met-api/src/met_api/resources/contact.py [hunk 3]
```diff
@@ -88,3 +88,26 @@ def get():
...
+    @staticmethod
+    # @TRACER.trace()
+    @cross_origin(origins=allowedorigins())
+    @auth.require
+    def put():
+        """Update Contact."""
+        try:
+            contact_data = TokenInfo.get_contact_data()
+            contact_schema = ContactSchema().load(user_data)
+            contact = ContactService().create_or_update_contact(contact_schema)
```
Comment (tiagograf-AOT) [inline met-api/src/met_api/resources/contact.py:107]: i think you want to use the update_contact method you created

File: met-api/src/met_api/resources/contact.py [hunk 4]
```diff
@@ -88,3 +88,26 @@ def get():
...
+    """Contact controller class."""
+
+    @staticmethod
+    # @TRACER.trace()
+    @cross_origin(origins=allowedorigins())
+    @auth.require
+    def put():
+        """Update Contact."""
+        try:
+            contact_data = TokenInfo.get_contact_data()
```
Comment (tiagograf-AOT) [inline met-api/src/met_api/resources/contact.py:105]: TokenInfo class is used to retrieve information about the logged in user, for your case you should use user_id = TokenInfo.get_id() to get the user id that is updating the contact, and pass the user_id along to you service class, there you'll need to update the updated_by in the contact record

File: met-api/src/met_api/models/contact.py [hunk 5]
```diff
@@ -58,3 +58,27 @@ def create_contact(cls, contact) -> Contact:
...
+
+        update_fields = dict(
+            name=contact_dict.get('name', contact.name),
+            title=contact_dict.get('title', contact.title),
+            phone_number=contact_dict.get('phone_number', contact.phone_number),
+            email=contact_dict.get('email', contact.email),
+            phone_number=contact_dict.get('phone_number', contact.phone_number),
+            address=contact_dict.get('address', contact.address),
+            bio=contact_dict.get('bio', contact.bio),
+            updated_date=datetime.utcnow(),
```
Comment (tiagograf-AOT) [inline met-api/src/met_api/models/contact.py:79]: also add the updated_by = user_id here

File: met-api/src/met_api/models/contact.py [hunk 6]
```diff
@@ -58,3 +58,27 @@ def create_contact(cls, contact) -> Contact:
...
         return new_contact
+    
+     
+    @classmethod
+    def update_contact(cls, contact_id, contact_dict) -> Optional[Contact]:
+        """Update contact."""
+        query = Contact.query.filter_by(id=contact_id)
+        contact: Contact = query.first()
+        if not contact:
+            return None
```
Comment (tiagograf-AOT) [inline met-api/src/met_api/models/contact.py:69]: if this is the case than something is wrong, we should throw an error saying the record was not found

File: met-api/src/met_api/resources/contact.py (hunk 2 above)
Comment (djnunez-aot) [reply met-api/src/met_api/resources/contact.py]: thank you for all the changes i addressed them

File: met-api/src/met_api/resources/contact.py (hunk 4 above)
Comment (djnunez-aot) [reply met-api/src/met_api/resources/contact.py]: awesome did not know this :)

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx [hunk 7]
```diff
@@ -17,40 +17,74 @@ import { WidgetDrawerContext } from './WidgetDrawerContext';
...
     });
 
+    useEffect(() => {
+        methods.setValue('name', selectedContact?.name);
+        methods.setValue('phone_number', selectedContact?.phone_number);
+        methods.setValue('title', selectedContact?.title);
+        methods.setValue('email', selectedContact?.email);
+        methods.setValue('address', selectedContact?.address);
+        methods.setValue('bio', selectedContact?.bio);
+    }, [selectedContact]);
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:54]: Is the defaultValues: { ... alone not working?

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx [hunk 8]
```diff
@@ -17,40 +17,74 @@ import { WidgetDrawerContext } from './WidgetDrawerContext';
...
+            !selectedContact
+                ? dispatch(
+                      openNotification({
+                          severity: 'error',
+                          text: 'An error occured while trying to create a new contact',
+                      }),
+                  )
+                : dispatch(
+                      openNotification({ severity: 'error', text: 'An error occured while trying to update contact' }),
+                  );
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:86]: Maybe one message and use the term 'save' contact for both

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx (hunk 7 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:54]: No it is not, not sure why

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx [hunk 9]
```diff
@@ -17,40 +17,74 @@ import { WidgetDrawerContext } from './WidgetDrawerContext';
...
-            dispatch(openNotification({ severity: 'success', text: 'A new contact was successfully added' }));
-            loadContacts();
+            if (selectedContact) {
+                // setIsCreatingContact(true);
+                // await updateContact(data);
+                // setIsCreatingContact(false);
+                // handleAddContactDrawerOpen(false);
+                // dispatch(openNotification({ severity: 'success', text: 'Contact was successfully updated' }));
+                // loadContacts();
+            } else {
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:67]: can we remove the if else here. Have one method in this component called saveContact. and there check go for :

if(selectedContact) {
     return updateContact
}
return postContact()

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx (hunk 8 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:86]: Ill try that

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx (hunk 9 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:67]: Perfect

File: met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx [hunk 10]
```diff
@@ -22,33 +113,51 @@ const ContantInfoPaper = ({ testId, contact, ...rest }: ContantInfoPaperProps) =
...
-                    spacing={2}
+                    spacing={1}
                 >
-                    <Grid item xs={2}>
-                        <MetParagraph>Phone:</MetParagraph>
-                    </Grid>
-                    <Grid item xs={10}>
-                        <MetParagraph>{contact.phone_number}</MetParagraph>
-                    </Grid>
+                    {contact.phone_number ? (
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx:123]: let's use <While> from react-if here

File: met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx [hunk 11]
```diff
@@ -22,33 +113,51 @@ const ContantInfoPaper = ({ testId, contact, ...rest }: ContantInfoPaperProps) =
...
                         <MetParagraph>{contact.email}</MetParagraph>
                     </Grid>
-
-                    <Grid item xs={2}>
-                        <MetParagraph>Address:</MetParagraph>
-                    </Grid>
-                    <Grid item xs={10}>
-                        <MetParagraph>{contact.address}</MetParagraph>
-                    </Grid>
+                    {contact.address !== '' ? (
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx:142]: let's use <While> from react-if here

File: met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx [hunk 12]
```diff
@@ -22,33 +113,51 @@ const ContantInfoPaper = ({ testId, contact, ...rest }: ContantInfoPaperProps) =
...
+                        <>
+                            <Grid item xs={2}>
+                                <MetParagraph>Address:</MetParagraph>
+                            </Grid>
+                            <Grid item xs={10}>
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx:147]: let's use <While> from react-if here

File: met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx [hunk 13]
```diff
@@ -106,7 +116,12 @@ export const WidgetDrawerProvider = ({ children }: { children: JSX.Element | JSX
         setWidgetDrawerOpen(open);
     };
 
-    const handleAddContactDrawerOpen = (open: boolean) => {
+    const handleAddContactDrawerOpen = (open: boolean, data?: Contact) => {
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx:119]: data -> contact

File: met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx [hunk 14]
```diff
@@ -106,7 +116,12 @@ export const WidgetDrawerProvider = ({ children }: { children: JSX.Element | JSX
...
+        if (data !== null) {
+            setSelectedContact(data);
+        } else {
+            setSelectedContact(null);
+        }
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx:124]: no if else, just setSelectedContact(data) if data is null it will set it to null then anyway

**Review by jadmsaadaot — CHANGES_REQUESTED** [review]
Added a few more comments.

Regarding style: let's try to minimize usage of if else blocks and ternaries. If we can find a cleaner way it's better because they make the code harder to change and they also cause duplication

File: met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx (hunk 14 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx:132]: thank you

File: met-web/src/App.tsx [hunk 15]
```diff
@@ -60,16 +62,21 @@ const App = () => {
...
         <Router>
-            <Box sx={{ display: 'flex' }}>
-                <LoggedInHeader drawerWidth={drawerWidth} />
-                <Notification />
-                <NotificationModal />
-                <Box component="main" sx={{ flexGrow: 1, width: `calc(100% - ${drawerWidth}px)`, marginTop: '17px' }}>
-                    <Toolbar />
-                    <AuthenticatedRoutes />
-                    <FeedbackModal />
+            <DndProvider backend={HTML5Backend}>
```
Comment (jadmsaadaot) [inline met-web/src/App.tsx:65]: if we needed providers can we add them with the other providers in the index.tsx file?

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx (hunk 7 above)
Comment (jadmsaadaot) [reply met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:55]: Okay well it's kinda weird that we need to do this. We need to look into it

File: met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx (hunk 8 above)
Comment (jadmsaadaot) [reply met-web/src/components/engagement/form/EngagementWidgets/AddContactDrawer.tsx:100]: reminding you of this

File: met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx [hunk 16]
```diff
@@ -1,55 +1,160 @@
...
-                    spacing={2}
+                    spacing={1}
                 >
-                    <Grid item xs={2}>
-                        <MetParagraph>Phone:</MetParagraph>
-                    </Grid>
-                    <Grid item xs={10}>
-                        <MetParagraph>{contact.phone_number}</MetParagraph>
-                    </Grid>
+                    <If condition={contact.phone_number}>
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx:126]: Not entirely sure, but If requires to have an <Else> if the condition was falsey. The <While> Tag doesn't, It just renders children if condition is truthy so maybe use that?

File: met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx [hunk 17]
```diff
@@ -8,15 +8,34 @@ import { postWidgetItems } from 'services/widgetService';
...
     const [addingWidgetItems, setAddingWidgetItems] = useState(false);
 
+    useEffect(() => {
+        const filteredContacts = contacts.filter((c: Contact) =>
+            addedContacts.map((contact) => contact.id).includes(c.id),
+        );
+        setAddedContacts(filteredContacts);
+    }, [contacts]);
+
+    const moveContact = useCallback((dragIndex: number, hoverIndex: number) => {
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx:28]: Question here, how are we saving the sorting order after the dragging happens? So like if contacts where A then B then C, user dragged B above A. When we open drawer again will we see B, A C?

File: met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx (hunk 13 above)
Comment (jadmsaadaot) [reply met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx:127]: reminding you of this

File: met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx [hunk 18]
```diff
@@ -44,16 +48,22 @@ export const WidgetDrawerContext = createContext<WidgetDrawerContextProps>({
...
+    updateSelectedContact: () => {
+        /* empty default method  */
+    },
 });
 
 export const WidgetDrawerProvider = ({ children }: { children: JSX.Element | JSX.Element[] }) => {
     const { savedEngagement } = useContext(ActionContext);
     const dispatch = useAppDispatch();
-
+    const [selectedContact, setSelectedContact] = useState<Contact | null | undefined>(null);
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/WidgetDrawerContext.tsx:66]: when is it undefined?

File: met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx [hunk 19]
```diff
@@ -32,6 +51,12 @@ const WhoIsListeningForm = () => {
         setAddedContacts([...addedContacts, selectedContact]);
     };
 
+    const removeContact = (contact: Contact) => {
+        const filteredContacts = addedContacts.filter((c) => contact.id != c.id);
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx:55]: https://github.com/ryanmcdermott/clean-code-javascript#variables

File: met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx [hunk 20]
```diff
@@ -8,15 +8,34 @@ import { postWidgetItems } from 'services/widgetService';
...
-    const { handleWidgetDrawerOpen, handleAddContactDrawerOpen, loadingContacts, contacts, widgets } =
+    const { handleWidgetDrawerOpen, handleAddContactDrawerOpen, loadingContacts, contacts, widgets, clearSelected } =
         useContext(WidgetDrawerContext);
     const dispatch = useAppDispatch();
     const [selectedContact, setSelectedContact] = useState<Contact | null>(null);
     const [addedContacts, setAddedContacts] = useState<Contact[]>([]);
     const [addingWidgetItems, setAddingWidgetItems] = useState(false);
 
+    useEffect(() => {
+        const filteredContacts = contacts.filter((c: Contact) =>
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx:22]: what's the logic here?

Also https://github.com/ryanmcdermott/clean-code-javascript#avoid-mental-mapping

File: met-api/src/met_api/models/contact.py (hunk 6 above)
Comment (jadmsaadaot) [reply met-api/src/met_api/models/contact.py:68]: Was this resloved?

File: met-web/src/components/common/ControlledInputComponents/ControlledTextField.tsx [hunk 21]
```diff
@@ -16,7 +16,7 @@ const ControlledTextField: FC<IFormInputProps> = ({ name, ...otherProps }) => {
         <Controller
             control={control}
             name={name}
-            defaultValue={defaultValues?.[name] || ''}
```
Comment (jadmsaadaot) [inline met-web/src/components/common/ControlledInputComponents/ControlledTextField.tsx:19]: why?

File: met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx [hunk 22]
```diff
@@ -1,55 +1,160 @@
...
+            // Note: we're mutating the monitor item here!
+            // Generally it's better to avoid mutations,
+            // but it's good here for the sake of performance
+            // to avoid expensive index searches.
+            item.index = hoverIndex;
+        },
+    });
+
+    const [{ isDragging }, drag] = useDrag({
+        type: 'Contact',
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx:88]: are we ignoring this warning?

**Review by jadmsaadaot — CHANGES_REQUESTED** [review]
A few logic and styling issues.

Also have we considered if it's possible to have a reusable drag and drop component. Is that possible?

File: met-api/src/met_api/models/contact.py (hunk 5 above)
Comment (jadmsaadaot) [reply met-api/src/met_api/models/contact.py]: not resolved

File: met-api/src/met_api/resources/contact.py [hunk 23]
```diff
@@ -88,3 +88,26 @@ def get():
...
+@API.route('/')
+class Contact(Resource):
+    """Contact controller class."""
+
+    @staticmethod
+    # @TRACER.trace()
+    @cross_origin(origins=allowedorigins())
+    @auth.require
+    def put():
+        """Update Contact."""
```
Comment (jadmsaadaot) [inline met-api/src/met_api/resources/contact.py]: Is this a Patch or Put?

File: met-api/src/met_api/resources/contact.py (hunk 4 above)
Comment (jadmsaadaot) [reply met-api/src/met_api/resources/contact.py]: Is this resolved @tiagograf-AOT ?

File: met-api/src/met_api/resources/contact.py (hunk 23 above)
Comment (djnunez-aot) [reply met-api/src/met_api/resources/contact.py]: That's patch on my version

File: met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx [hunk 24]
```diff
@@ -1,64 +1,195 @@
...
+                    spacing={1}
                 >
-                    <Grid item xs={2}>
-                        <MetParagraph>Phone:</MetParagraph>
-                    </Grid>
-                    <Grid item xs={10}>
-                        <MetParagraph>{contact.phone_number}</MetParagraph>
-                    </Grid>
+                    <If condition={!!contact.phone_number}>
+                        <Then>
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx:127]: please replace if with <While>

File: met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx (hunk 17 above)
Comment (saravanpa-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/WhoIsListeningForm.tsx:28]: sorting order is important ; each contact has to be shown in the order they were dragged... I am okay to handle it in next PR

**Review by saravanpa-aot — APPROVED** [review]

File: met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx (hunk 16 above)
Comment (jadmsaadaot) [reply met-web/src/components/engagement/form/EngagementWidgets/ContactInfoPaper.tsx]: The While* tag

**Review by jadmsaadaot — APPROVED** [review]
